From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Faith Ekstrand <faith.ekstrand@collabora.com>
Date: Tue, 18 Nov 2025 10:02:29 -0500
Subject: [PATCH 04/57] panvk/csf: Bind GS/TES/TCS descriptors

Unlike FS/VS, these don't have any funny workarounds or extra bits we
need to tuck into the descriptors.  They just have the client-bound
descriptor sets.

(cherry picked from commit 876de9a6763ef65cacdfa1bf41eca4f830100b3b)
---
 src/panfrost/vulkan/csf/panvk_vX_cmd_draw.c | 113 ++++++++++++++++++++
 1 file changed, 113 insertions(+)

diff --git a/src/panfrost/vulkan/csf/panvk_vX_cmd_draw.c b/src/panfrost/vulkan/csf/panvk_vX_cmd_draw.c
index a043e61a57a..a0ccc9b2d1f 100644
--- a/src/panfrost/vulkan/csf/panvk_vX_cmd_draw.c
+++ b/src/panfrost/vulkan/csf/panvk_vX_cmd_draw.c
@@ -457,6 +457,98 @@ prepare_vs_desc(struct panvk_cmd_buffer *cmdbuf,
    return VK_SUCCESS;
 }
 
+static VkResult
+prepare_shader_driver_set(struct panvk_cmd_buffer *cmdbuf,
+                          const struct panvk_shader *shader,
+                          struct panvk_shader_desc_state *shader_desc_state)
+{
+   const struct panvk_shader_desc_info *desc_info = &shader->desc_info;
+   uint32_t desc_count = desc_info->dyn_bufs.count + 1;
+
+   struct pan_ptr driver_set = panvk_cmd_alloc_dev_mem(
+      cmdbuf, desc, desc_count * PANVK_DESCRIPTOR_SIZE, PANVK_DESCRIPTOR_SIZE);
+   struct panvk_opaque_desc *descs = driver_set.cpu;
+
+   if (desc_count && !driver_set.gpu)
+      return VK_ERROR_OUT_OF_DEVICE_MEMORY;
+
+   /* Dummy sampler always comes first */
+   pan_cast_and_pack(&descs[0], SAMPLER, cfg) {
+      cfg.clamp_integer_array_indices = false;
+   }
+
+   const struct panvk_descriptor_state *desc_state =
+      &cmdbuf->state.gfx.desc_state;
+
+   panvk_per_arch(cmd_fill_dyn_bufs)(
+      desc_state, desc_info, (struct mali_buffer_packed *)(&descs[1]));
+
+   shader_desc_state->driver_set.dev_addr = driver_set.gpu;
+   shader_desc_state->driver_set.size = desc_count * PANVK_DESCRIPTOR_SIZE;
+   gfx_state_set_dirty(cmdbuf, DESC_STATE);
+   return VK_SUCCESS;
+}
+
+static VkResult
+prepare_shader_desc(struct panvk_cmd_buffer *cmdbuf,
+                    const struct panvk_shader *shader,
+                    struct panvk_shader_desc_state *shader_desc_state)
+{
+   if (!shader) {
+      memset(shader_desc_state, 0, sizeof(*shader_desc_state));
+      return VK_SUCCESS;
+   }
+
+   VkResult result = prepare_shader_driver_set(cmdbuf, shader,
+                                               shader_desc_state);
+   if (result != VK_SUCCESS)
+      return result;
+
+   const struct panvk_shader_desc_info *desc_info = &shader->desc_info;
+   const struct panvk_descriptor_state *desc_state =
+      &cmdbuf->state.gfx.desc_state;
+
+   result = panvk_per_arch(cmd_prepare_shader_res_table)(
+      cmdbuf, desc_state, desc_info, shader_desc_state, 1);
+   if (result != VK_SUCCESS)
+      return result;
+
+   return VK_SUCCESS;
+}
+
+static VkResult
+prepare_gs_desc(struct panvk_cmd_buffer *cmdbuf)
+{
+   if (!gfx_state_dirty(cmdbuf, GS) &&
+       !gfx_state_dirty(cmdbuf, DESC_STATE))
+      return VK_SUCCESS;
+
+   return prepare_shader_desc(cmdbuf, cmdbuf->state.gfx.gs.shader,
+                              &cmdbuf->state.gfx.gs.desc);
+}
+
+static VkResult
+prepare_tcs_desc(struct panvk_cmd_buffer *cmdbuf)
+{
+   if (!gfx_state_dirty(cmdbuf, TCS) &&
+       !gfx_state_dirty(cmdbuf, DESC_STATE))
+      return VK_SUCCESS;
+
+   return prepare_shader_desc(cmdbuf, cmdbuf->state.gfx.tcs.shader,
+                              &cmdbuf->state.gfx.tcs.desc);
+}
+
+static VkResult
+prepare_tes_desc(struct panvk_cmd_buffer *cmdbuf)
+{
+   if (!gfx_state_dirty(cmdbuf, TES) &&
+       !gfx_state_dirty(cmdbuf, DESC_STATE))
+      return VK_SUCCESS;
+
+   return prepare_shader_desc(cmdbuf, cmdbuf->state.gfx.tes.shader,
+                              &cmdbuf->state.gfx.tes.desc);
+}
+
 static void
 emit_varying_descs(const struct panvk_cmd_buffer *cmdbuf,
                    struct mali_attribute_packed *descs)
@@ -588,6 +680,9 @@ prepare_descs(struct panvk_cmd_buffer *cmdbuf,
               const struct panvk_draw_info *draw)
 {
    const struct panvk_shader *vs = cmdbuf->state.gfx.vs.shader;
+   const struct panvk_shader *tcs = cmdbuf->state.gfx.tcs.shader;
+   const struct panvk_shader *tes = cmdbuf->state.gfx.tes.shader;
+   const struct panvk_shader *gs = cmdbuf->state.gfx.gs.shader;
    const struct panvk_shader *fs = get_fs(cmdbuf);
    struct panvk_descriptor_state *desc_state =
       &cmdbuf->state.gfx.desc_state;
@@ -595,8 +690,14 @@ prepare_descs(struct panvk_cmd_buffer *cmdbuf,
 
    if (gfx_state_dirty(cmdbuf, DESC_STATE) ||
        gfx_state_dirty(cmdbuf, VS) ||
+       gfx_state_dirty(cmdbuf, TCS) ||
+       gfx_state_dirty(cmdbuf, TES) ||
+       gfx_state_dirty(cmdbuf, GS) ||
        fs_user_dirty(cmdbuf)) {
       uint32_t used_set_mask = vs->desc_info.used_set_mask;
+      used_set_mask |= gs ? gs->desc_info.used_set_mask : 0;
+      used_set_mask |= tcs ? tcs->desc_info.used_set_mask : 0;
+      used_set_mask |= tes ? tes->desc_info.used_set_mask : 0;
       used_set_mask |= fs ? fs->desc_info.used_set_mask : 0;
 
       result = panvk_per_arch(cmd_prepare_push_descs)(cmdbuf, desc_state,
@@ -609,6 +710,18 @@ prepare_descs(struct panvk_cmd_buffer *cmdbuf,
    if (result != VK_SUCCESS)
       return result;
 
+   result = prepare_tes_desc(cmdbuf);
+   if (result != VK_SUCCESS)
+      return result;
+
+   result = prepare_tcs_desc(cmdbuf);
+   if (result != VK_SUCCESS)
+      return result;
+
+   result = prepare_gs_desc(cmdbuf);
+   if (result != VK_SUCCESS)
+      return result;
+
    result = prepare_fs_desc(cmdbuf);
    if (result != VK_SUCCESS)
       return result;
